stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb

```diff
@@ -0,0 +1,275 @@
+          address._rawValue, desired.${getter}))
+%   end
+    }
+  }
+
+  /// Perform an atomic compare and exchange operation with
+  /// the specified memory ordering.
+  ///
+  /// This operation is equivalent to the following pseudocode:
+  ///
+  /// ```
+  /// atomic(self, ordering: ordering) { value in
+  ///   guard value == expected else { return (false, value) }
+  ///   value = desired
+  ///   return (true, expected)
+  /// }
+  /// ```
+  ///
+  /// This method implements a "strong" compare and exchange operation
+  /// that does not permit spurious failures.
+  @_semantics("has_constant_evaluable_arguments")
+  @_transparent @_alwaysEmitIntoClient
+  public static func atomicCompareExchange(
+    expected: ${swiftType},
+    desired: ${swiftType},
+    at address: UnsafeMutablePointer<${swiftType}>,
+    ordering: AtomicUpdateOrdering
+  ) -> (exchanged: Bool, original: ${swiftType}) {
+    switch ordering {
+%   for (swiftOrder, llvmOrder, failureOrder) in rmwOrderings():
+    ${caseStatementForOrdering(swiftOrder)}:
+      let (oldValue, won) = Builtin.cmpxchg_${llvmOrder}_${failureOrder}_${builtinType}(
+        address._rawValue, expected.${getter}, desired.${getter})
+      return (Bool(_builtinBooleanLiteral: won), ${swiftType}(oldValue))
+%   end
+    }
+  }
+
+  /// Perform an atomic compare and exchange operation with the specified
+  /// success/failure memory orderings.
+  ///
+  /// This operation is equivalent to the following pseudocode:
+  ///
+  /// ```
+  /// atomic(self) { currentValue in
+  ///   let original = currentValue
+  ///   guard original == expected else { return (false, original) }
+  ///   currentValue = desired
+  ///   return (true, original)
+  /// }
+  /// ```
+  ///
+  /// The `ordering` argument specifies the memory ordering to use when the
+  /// operation manages to update the current value, while `failureOrdering`
+  /// will be used when the operation leaves the value intact.
+  ///
+  /// The `failureOrdering` argument currently isn't allowed to introduce a
+  /// synchronization constraint that isn't also guaranteed by `ordering`.
+  /// (This limitation may be lifted in the future.) For example:
+  ///
+  /// ```
+  /// // OK
+  /// ptr._atomicCompareExchangeWord(
+  ///   expected: 1, desired: 2,
+  ///   ordering: .acquiringAndReleasing, failureOrdering: .relaxed)
+  ///
+  /// // Not supported:
+  /// ptr._atomicCompareExchangeWord(
+  ///   expected: 3, desired: 0,
+  ///   ordering: .releasing, failureOrdering: .acquiring)
+  /// ```
+  @_semantics("has_constant_evaluable_arguments")
+  @_transparent @_alwaysEmitIntoClient
+  public static func atomicCompareExchange(
+    expected: ${swiftType},
+    desired: ${swiftType},
+    at address: UnsafeMutablePointer<${swiftType}>,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: ${swiftType}) {
+    switch (ordering, failureOrdering) {
```

Comment:
Is this one also a strong CAS?

Reply:
Ah, yes, it is! I'll need to do another pass on updating the docs.

Reply:
Done; I also added a separate weak variant (two orderings only, for now).